### PR TITLE
Attempt to convert service worker to TS

### PR DESCRIPTION
**Pull request by glennhartmann**
(depends on https://github.com/GoogleChromeLabs/llaminator/pull/14)

`npx tsc --lib webworker src/sw.ts` builds this without error (although `npx tsc src/sw.ts` does not, which makes me question what exactly the `tsconfig.json` does).

But I get this extremely helpful runtime error when using `npm run-script serve`:

```
Uncaught SyntaxError: Cannot use import statement outside a module
Uncaught (in promise) TypeError: Failed to register a ServiceWorker for scope ('http://localhost:4629/src/') with script ('http://localhost:4629/src/sw.ts'): ServiceWorker script evaluation failed
```

Any ideas? Shouldn't ESBuild be taking care of this?

(This is what I mean when I say that setting up typescript and a build system is a massive mess)

### CODE REVIEW

**Comment by glennhartmann** [comment]
on subsequent attempts, it doesn't even mention `Uncaught SyntaxError: Cannot use import statement outside a module`, it just says

```
Uncaught (in promise) TypeError: Failed to register a ServiceWorker for scope ('http://localhost:4629/src/') with script ('http://localhost:4629/src/sw.ts'): ServiceWorker script evaluation failed
```

File: tsconfig.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
     "moduleResolution": "node",
     "esModuleInterop": false,
     "noEmitOnError": true,
-    "lib": ["esnext", "dom", "dom.iterable"],
+    "lib": ["esnext", "dom", "dom.iterable", "webworker"],
```
Comment (beverloo) [inline tsconfig.json:8]: This is tricky—we want the `webworker` library to be available in the service worker, but not in other code. Conversely, the `dom` and `dom.iterable` libraries should not be available in the service worker. Jake has a repository w/ the following example, where includes are used in `tsconfig.json` to enable different environments:

https://github.com/jakearchibald/typescript-worker-example

Once the commands work, we can codify them in `package.json`.

File: package.json [hunk 2]
```diff
@@ -14,5 +14,9 @@
     "@web/dev-server": "^0.1.28",
     "@web/dev-server-esbuild": "^0.2.16",
     "typescript": "^4.5.2"
+  },
+  "dependencies": {
+    "workbox-routing": "^6.4.1",
+    "workbox-strategies": "^6.4.1"
```
Comment (beverloo) [inline package.json:20]: We should probably think about whether we want to use workbox or not. I'm fine either way, but I don't think it gives us too much. (SWR is easy enough to implement, and we don't have sophisticated logic.)

File: src/sw.ts [hunk 3]
```diff
@@ -14,40 +14,43 @@
...
+import { StaleWhileRevalidate } from 'workbox-strategies';
 
 const cacheName = 'app-cache';
 
 const mainDBName = 'appDB';
 const objStoreName = 'imageStore';
 const mainImageName = 'mainImage';
 
-self.addEventListener('install', function (event) {
+interface InstallEvent extends Event {
```
Comment (beverloo) [inline src/sw.ts:26]: Declaring `self` as the right interface will probably remove the need for this:

```
declare var self: ServiceWorkerGlobalScope;
```

**Review by beverloo — APPROVED** [review]
LG, but without the build steps this will break. (However, a bit of breakage is fine to remove the chain of dependencies we're in IMO.)

File: src/sw.ts (hunk 3 above)
Comment (glennhartmann) [reply src/sw.ts]: done

**Pull request merged**